### PR TITLE
Autofixing: initial implementation

### PR DESCRIPTION
**Pull request by mdevils**
**Work in progress**

In this pull-request I have implemented autofixing mechanism:

* Flag `-x, --fix` does autofixing.
* In case of `stdin` input, autofixed content is written to the console.
* I tried to keep everything as simple as possible.
* I will write tests later.
* I will cover other assertion methods soon.
* All unfixed errors are printed.

@mikesherov @markelog @mrjoelkemp @zxqfox please take a look!

### CODE REVIEW

**Comment by hzoo** [comment]
Wow this is awesome! I just tried a simple test

```js
// var test = function(options){
//     for (var i=0;i<asdf.length;i++)
//         asdf();
//     x+ y;
// }

var test = function(options) {
    for (var i = 0; i < asdf.length; i++)
        asdf();
    x + y;
}
```

Fixes all whitespace errors and prints unfixed ones (indentation, curly braces)

File: lib/cli.js [hunk 1]
```diff
@@ -114,14 +114,24 @@ module.exports = function(program) {
...
-            return [errors];
-        });
+        if (program.fix) {
+            return {
+                promise: checker.fixStdin().then(function(result) {
+                    process.stdout.write(result.output);
+                }),
+                checker: checker
+            };
+        } else {
```
Comment (zxqfox) [inline lib/cli.js]: If you returning in `if`, why to use `else`?

File: lib/token-assert.js [hunk 2]
```diff
@@ -31,19 +31,23 @@ TokenAssert.prototype.whitespaceBetween = function(options) {
         if (nextToken.loc.start.line === token.loc.end.line &&
             (nextToken.loc.start.column - token.loc.end.column) !== spaces
         ) {
+            nextToken.whitespaceBefore = new Array(spaces + 1).join(' ');
```
Comment (zxqfox) [inline lib/token-assert.js]: Is it right that we fixing code right in assert methods?

File: lib/string-checker.js [hunk 3]
```diff
@@ -136,8 +150,87 @@ StringChecker.prototype = {
...
+                        break;
+                    }
+                }
+
+                if (!hasFixes) {
+                    break;
+                }
+
+                source = file.render();
+                sourceTree = JsFile.parse(source, this._esprima, this._configuration.getEsprimaOptions());
```
Comment (zxqfox) [inline lib/string-checker.js]: What is the case to render/parsing multiple time?

**Comment by zxqfox** [comment]
Looks awesome!

File: lib/string-checker.js (hunk 3 above)
Comment (mikesherov) [inline lib/string-checker.js]: That the fix for one rule causes a violation of another.

File: lib/string-checker.js [hunk 4]
```diff
@@ -136,8 +150,87 @@ StringChecker.prototype = {
...
+
+                attempt++;
+            } while (attempt < MAX_FIX_ATTEMPTS);
```
Comment (mikesherov) [inline lib/string-checker.js]: Doing 5 attempts is suboptimal. We can likely do the following instead:

1. Instead of counting the number of attempts we made, count the number of violations remaining.

2. Keep looping as long as `fixes > 0` and `violations < violationsOfPreviousRun`. 

This ensures that we keep fixing until it's no longer helping, which will break a cycle after 1 iteration, not 5, and in the rare case where you'd have to do more than 5 iterations, it'll keep going until it's not helping.

File: lib/token-assert.js (hunk 2 above)
Comment (mikesherov) [inline lib/token-assert.js]: Yes, but this modification should likely only happen if fixing is turned on.

Eventually, rules might actually be checking the whitespaceBefore property, and if we unconditionally change this whitespace, than overlapping rules will fail to report an error at the same location if we're not in fix mode.

File: lib/cli.js (hunk 1 above)
Comment (mdevils) [inline lib/cli.js]: Correct structures are easy to read.

File: lib/token-assert.js (hunk 2 above)
Comment (mdevils) [inline lib/token-assert.js]: I believe rules should not access `whitespaceBefore` property.

File: lib/string-checker.js (hunk 4 above)
Comment (mdevils) [inline lib/string-checker.js]: > Keep looping as long as fixes > 0 and violations < violationsOfPreviousRun.

This might not work. For instance one rule does 1 fix and then there are 2 or more fixes needs to be done also.

File: lib/string-checker.js (hunk 3 above)
Comment (zxqfox) [inline lib/string-checker.js]: But... I believe that we shouldn't allow checking with conflicting rules like requireSomething and disallowSomething at once. And at that POV this line looks like overkill.

File: lib/string-checker.js (hunk 4 above)
Comment (mikesherov) [inline lib/string-checker.js]: You mean 1 fix increases violations for next run? I see.

What about the case where more than 5 runs are needed? I doubt such a case exists, so this is fine for now I suppose.

File: lib/string-checker.js (hunk 3 above)
Comment (mikesherov) [inline lib/string-checker.js]: Not conflicting rules, complimentary rules. For example, our spacing rules typically ignore tokens that are on the next line.

If a "disallowNewLine" fix moves the token onto the same line as the previous token, it may now be in violation of a "disallowSpace" rule.

File: lib/cli.js (hunk 1 above)
Comment (zxqfox) [inline lib/cli.js]: Incorrect structure are hard to parse ;-) Use of `return` makes an usual `if-else` harder to read. That's why I've asked ;-)

File: lib/token-assert.js (hunk 2 above)
Comment (zxqfox) [inline lib/token-assert.js]: Yeah, +1 @mdevils, +1 @mikesherov for only if fixing is on.

Comment (mikesherov) [inline lib/token-assert.js]: Well, think of it this way. If we both modify whitespaceBefore and use it for checking, then we avoid a lot of cases where two full runs would be needed.

Imagine a "requireNewline" rule and a "requireSpace" rule that both report on a `{` at the end of a line...

If we had rules check the whitespaceBefore property, then a fix for requireNewline would make requireSpace ignore.

If we don't, then 2 passes must happen, one fixing the spacing, while the second pass finally moved it to the right spot on the next line.

Comment (mikesherov) [inline lib/token-assert.js]: Anyway, we can defer this discussion till later, especially consider this is all consolidated to one spot considering we have this nice assertion layer.

File: lib/string-checker.js (hunk 4 above)
Comment (zxqfox) [inline lib/string-checker.js]: Feels like we need some specs for this ;-) for objective discussing and to get rid of "might".

File: lib/string-checker.js (hunk 3 above)
Comment (zxqfox) [inline lib/string-checker.js]: Oh, right. Thanks. But should we really reparse? Probably there is a way to just recheck fixed tree. It's much harder, yes. Ok, nvm ;-(

Comment (mdevils) [inline lib/string-checker.js]: Fixing token without reparse it way harder.

File: lib/token-assert.js (hunk 2 above)
Comment (mdevils) [inline lib/token-assert.js]: Yeah, I vote for deferring this discussion till later. I believe good assertion layer must cover 100% of cases.

File: lib/cli.js (hunk 1 above)
Comment (mdevils) [inline lib/cli.js]: OK, I will fix this.

File: lib/token-assert.js (hunk 2 above)
Comment (zxqfox) [inline lib/token-assert.js]: Agreed totally.

File: lib/string-checker.js (hunk 4 above)
Comment (mdevils) [inline lib/string-checker.js]: I can't say I like `5` runs approach, but I think we can leave this for now and watch for user feedback.

Comment (mikesherov) [inline lib/string-checker.js]: @mdevils can you file an issue that we should revisit this later? I don't want it holding up the PR, but I don't want to lose track of the action item.

**Comment by mdevils** [comment]
I think this can be merged now :)

**Comment by mrjoelkemp** [comment]
Taking a look

File: lib/string-checker.js (hunk 4 above)
Comment (mdevils) [inline lib/string-checker.js]: Here we go: #1062

File: lib/checker.js [hunk 5]
```diff
@@ -45,12 +45,62 @@ Checker.prototype.checkFile = function(path) {
 };
 
 /**
+ * Fixes single file.
+ *
+ * @param {String} path
+ * @returns {Promise * Errors}
+ */
+Checker.prototype.fixFile = function(path) {
+    if (!this._configuration.isFileExcluded(path)) {
```
Comment (mrjoelkemp) [inline lib/checker.js]: Non-blocker: This style becomes pervasive throughout community contributions. What about opting for an early return?

```js
if (this._configuration.isFileExcluded(path)) {
  return Vow.resolve(null);
}

return vowFs.read(path, ...
```

The idea is along the lines of: http://sourcemaking.com/refactoring/replace-nested-conditional-with-guard-clauses

File: lib/string-checker.js [hunk 6]
```diff
@@ -77,38 +79,50 @@ StringChecker.prototype = {
...
+     * Fills Errors instance with validation errors.
+     *
+     * @param {JsFile} file
+     * @param {Errors} errors
+     * @private
+     */
+    _checkJsFile: function(file, errors) {
+        var errorFilter = this._configuration.getErrorFilter();
+
+        if (!this._maxErrorsExceeded) {
```
Comment (mikesherov) [inline lib/string-checker.js]: perhaps make this an early return now that it's an `if` enclosing almost the entire body of the function?

File: lib/checker.js [hunk 7]
```diff
@@ -45,12 +45,62 @@ Checker.prototype.checkFile = function(path) {
...
  */
 Checker.prototype.checkDirectory = function(path) {
+    return this._processDirectory(path, this.checkFile.bind(this));
+};
+
+/**
+ * Checks directory or file.
+ *
+ * @param {String} path
+ * @returns {Promise * Error[]}
```
Comment (mrjoelkemp) [inline lib/checker.js]: Non-blocker: I was wondering where this came from in https://github.com/jscs-dev/node-jscs/pull/1038/files#diff-605f23c7e3628b173a428e6a469e6ff6R52. 

Is that valid jsdoc?

File: lib/string-checker.js [hunk 8]
```diff
@@ -136,8 +150,87 @@ StringChecker.prototype = {
...
+            var errors;
+            var file;
+
+            do {
+                file = this._createJsFileInstance(filename, source, sourceTree);
+                errors = new Errors(file, this._verbose);
+                this._checkJsFile(file, errors);
+                var hasFixes = false;
+                var errorList = errors.getErrorList();
+                for (var i = 0; i < errorList.length; i++) {
```
Comment (mikesherov) [inline lib/string-checker.js]: I would encapsulate this as a method in error object: `hasFixes`. The `break` inside the code here is confusing because it breaks out of the `for` but not the `do`.

File: lib/string-checker.js [hunk 9]
```diff
@@ -136,8 +150,87 @@ StringChecker.prototype = {
...
+            es3: this._configuration.isES3Enabled(),
+            es6: this._configuration.isESNextEnabled()
+        });
+    },
+
+    /**
+     * Checks file provided with a string.
+     * @param {String} source
+     * @param {String} [filename]
+     * @returns {{output: String, errors: Errors}}
```
Comment (mrjoelkemp) [inline lib/string-checker.js]: I thought the jsdoc way to do this was:

```js
@returns {Object} data
@returns {String} data.output
@returns {Errors} data.errors
```

cc @zxqfox

File: lib/string-checker.js [hunk 10]
```diff
@@ -136,8 +150,87 @@ StringChecker.prototype = {
...
+        return new JsFile(filename, source, sourceTree, {
+            es3: this._configuration.isES3Enabled(),
+            es6: this._configuration.isESNextEnabled()
+        });
+    },
+
+    /**
+     * Checks file provided with a string.
+     * @param {String} source
+     * @param {String} [filename]
```
Comment (mrjoelkemp) [inline lib/string-checker.js]: `[filename='input']`

File: lib/string-checker.js (hunk 9 above)
Comment (mdevils) [inline lib/string-checker.js]: This is a valid JsDoc: http://usejsdoc.org/tags-type.html

Comment (mrjoelkemp) [inline lib/string-checker.js]: :+1:

File: lib/checker.js (hunk 7 above)
Comment (mdevils) [inline lib/checker.js]: This case is partially covered in JsDoc, so different projects use different notations: `{Promise ~ Type}`, others — `{Promise * Type}`. Following jsdoc documentation, the only close notation is `{Promise<Type>}`.

Comment (mrjoelkemp) [inline lib/checker.js]: :+1:

File: lib/string-checker.js (hunk 8 above)
Comment (mrjoelkemp) [inline lib/string-checker.js]:
```js
var hasFixes = errorList.some(function(err) {
  return err.fixed;
})
```

File: lib/checker.js (hunk 7 above)
Comment (zxqfox) [inline lib/checker.js]: Nope. https://github.com/jsdoc3/jsdoc/issues/509
The most logical for me is to use something like `{Promise.<Object,Error>}` or `{Promise}`

File: lib/string-checker.js [hunk 11]
```diff
@@ -136,8 +150,87 @@ StringChecker.prototype = {
...
+        if (parseError) {
+            var parseErrors = new Errors(this._createJsFileInstance(filename, source, sourceTree), this._verbose);
+            this._addParseError(parseErrors, parseError);
+            return {output: source, errors: parseErrors};
+        } else {
+            var attempt = 0;
+            var errors;
+            var file;
+
+            do {
```
Comment (mrjoelkemp) [inline lib/string-checker.js]: What in this `do` actually fixes the source? I see no explicit call to any fix-related function.

File: lib/string-checker.js (hunk 8 above)
Comment (mdevils) [inline lib/string-checker.js]: This looks easier, thanks

File: lib/string-checker.js (hunk 11 above)
Comment (mdevils) [inline lib/string-checker.js]: Assertions.

Comment (mikesherov) [inline lib/string-checker.js]: @mrjoelkemp checking the file automatically fixes the whitespace, which then is realized by rendering the file.

File: lib/string-checker.js [hunk 12]
```diff
@@ -136,8 +150,87 @@ StringChecker.prototype = {
...
+                        hasFixes = true;
+                        break;
+                    }
+                }
+
+                if (!hasFixes) {
+                    break;
+                }
+
+                source = file.render();
```
Comment (mikesherov) [inline lib/string-checker.js]: @mrjoelkemp here is where the fixes are reapplied back to the source, as a result of a parse on the next line.

File: lib/string-checker.js (hunk 9 above)
Comment (zxqfox) [inline lib/string-checker.js]: Btw you can look at https://developers.google.com/closure/compiler/docs/js-for-compiler there are a lot of different examples for many cases.

File: lib/string-checker.js (hunk 11 above)
Comment (mrjoelkemp) [inline lib/string-checker.js]: Thanks.

File: lib/string-checker.js [hunk 13]
```diff
@@ -136,8 +150,87 @@ StringChecker.prototype = {
...
+            return {output: source, errors: parseErrors};
+        } else {
+            var attempt = 0;
+            var errors;
+            var file;
+
+            do {
+                file = this._createJsFileInstance(filename, source, sourceTree);
+                errors = new Errors(file, this._verbose);
+                this._checkJsFile(file, errors);
```
Comment (mrjoelkemp) [inline lib/string-checker.js]: Non-blocker: consider leaving a comment above that indicates the implicit assertions fix. Something like Mike's explanation would be helpful to newcomers to the source who are not familiar with the inner assertions.

File: test/string-checker.js [hunk 14]
```diff
@@ -297,29 +339,20 @@ describe('modules/string-checker', function() {
...
-            checker = new Checker();
+            checker = new StringChecker();
             checker.registerDefaultRules();
         });
 
         it('should accept a path to a filter function to filter out errors', function() {
             checker.configure({
                 disallowQuotedKeysInObjects: true,
-                errorFilter: __dirname + '/data/error-filter.js'
+                errorFilter: require(__dirname + '/data/error-filter.js')
```
Comment (mrjoelkemp) [inline test/string-checker.js]: Why?

Comment (mrjoelkemp) [inline test/string-checker.js]: This is the only change that blocks this PR for me. Does this require a change in the way users use this feature?

Comment (mdevils) [inline test/string-checker.js]: This is `StringChecker` test and for some reason `StringChecker` was not actually tested here. `Checker` was. I have changed `new Checker(...)` to `new StringChecker(...)` and it required to change some tests for `StringChecker` compatibility.

> Does this require a change in the way users use this feature?

Nope.

File: lib/string-checker.js (hunk 13 above)
Comment (mdevils) [inline lib/string-checker.js]: I agree.

File: test/string-checker.js (hunk 14 above)
Comment (mrjoelkemp) [inline test/string-checker.js]: :+1:

**Comment by mrjoelkemp** [comment]
LGTM. Great work Marat! This is really exciting.

**Comment by mdevils** [comment]
Thank you guys! I will make fixes you have suggested and will merge this soon.

File: lib/string-checker.js (hunk 4 above)
Comment (gero3) [inline lib/string-checker.js]: 5 is just about right I think. I already had runs that took 3 times to fix it and that took 3 different rules in my autofixer. But mine doesn't fix the string at the same time of checking it, so this will need a lot of less runs.

File: lib/string-checker.js (hunk 6 above)
Comment (mdevils) [inline lib/string-checker.js]: I agree.

File: lib/checker.js (hunk 5 above)
Comment (mdevils) [inline lib/checker.js]: I agree.

File: lib/string-checker.js (hunk 4 above)
Comment (mikesherov) [inline lib/string-checker.js]: @gero3, thanks for the insight. And thanks for the initial nudge. Even though we didn't end up going the "format function" route, your initial PR was invaluable. Thanks again!

File: lib/checker.js [hunk 15]
```diff
@@ -35,13 +35,32 @@ Checker.prototype.configure = function(/*config*/) {
...
+ */
+Checker.prototype.fixFile = function(path) {
+    if (this._configuration.isFileExcluded(path)) {
+        return Vow.resolve(null);
+    }
+
+    return vowFs.read(path, 'utf8').then(function(data) {
+        var result = this.fixString(data, path);
+        return vowFs.write(path, result.output).then(function() {
+            return result.errors;
```
Comment (mikesherov) [inline lib/checker.js]: Ah, @mdevils, here is where it writes files. Sorry I missed it!
